src/app/core/common-components/entity-select/entity-select.component.ts

```diff
@@ -139,86 +128,87 @@ export class EntitySelectComponent<
+      }
 
-    await this.alignAvailableAndSelectedEntities(newAvailableEntities);
-
-    this.availableOptions.next(newAvailableEntities);
-    this.recalculateMatchingInactive();
-  }
+      return entities
+        .filter((e) => this.additionalFilter(e))
+        .sort((a, b) => a.toString().localeCompare(b.toString()));
+    },
+  });
 
   /**
-   * Edit form value (currently selected) and the given available Entities to be consistent:
-   * Entities that do not exist should be removed from the form value
-   * and availableEntities should contain all selected entities, even from other types.
-   * @private
+   * true when this is loading and false when it's ready.
+   * This subject's state reflects the actual loading resp. the 'readiness'-
+   * state of this component. Will trigger once loading is done
    */
-  private async alignAvailableAndSelectedEntities(availableEntities: E[]) {
-    if (this.form?.value === null || this.form?.value === undefined) {
-      return;
-    }
-
-    let updatedValue: T = this.form.value;
+  loading: Signal<boolean> = computed(() => this.entitiesForType.isLoading());
 
-    for (const id of asArray(this.form.value)) {
-      // Skip null, undefined, or empty string values
-      if (id === null || id === undefined || id === "") {
-        continue;
-      }
+  /**
+   * The currently selected values (IDs) of the form control.
+   */
+  values: Signal<string[]> = toSignal(
+    toObservable(this.form)
+      .pipe(switchMap((form) => form.valueChanges))
+      .pipe(map(asArray)),
+    { initialValue: [] },
+  );
+
+  private availableOptionsResource: Resource<E[]> = resource({
+    defaultValue: [],
+    params: () => ({
+      allEntities: this.entitiesForType.value(),
+      values: this.values(),
+      includeInactive: this.includeInactive(),
+    }),
+    loader: async ({ params }) => {
+      const includeSelected = (entity: E) =>
+        asArray(params.values).includes(entity.getId());
+
+      const availableEntities = params.allEntities.filter(
+        (e) => params.includeInactive || e.isActive || includeSelected(e),
+      );
 
-      if (availableEntities.find((e) => id === e.getId())) {
-        // already available, nothing to do
-        continue;
+      if (params.values !== null && params.values !== undefined) {
+        for (const id of asArray(params.values)) {
+          if (id === null || id === undefined || id === "") {
+            continue;
+          }
+
+          if (availableEntities.find((e) => id === e.getId())) {
+            continue;
+          }
+
+          const additionalEntity = await this.getEntity(id);
+          if (additionalEntity) {
+            availableEntities.push(additionalEntity);
+          } else {
+            this.hasInaccessibleEntities = true;
+            availableEntities.push({
+              getId: () => id,
+              isHidden: true,
+            } as unknown as E);
+          }
```

Comment:
maybe extract into private helper method to reduce complexity of this loader function?

Reply:
Some of the other changes reduced the complexity of the function. To me it doesn’t make sense to split up the function because it’s logic belongs together. And creating a separate function would just create some indirection. Let me know if you still would like it to be extracted.